### PR TITLE
feat(animations-moti): zero-render pseudo animations

### PR DESCRIPTION
**Pull request by nandorojo**
This PR adds animations powered by Moti for pseudo states, without firing renders. This PR moves press/hover/focus animation state onto the UI thread using Moti + Shared Values from Reanimated.

https://discord.com/channels/909986013848412191/1093629818253344808/1136770882723922081

Components are still double-rendering on mount (which shouldn't be needed with Moti) but we can address this in a separate PR.

Let me know if you ahve questions.

### CODE REVIEW

File: packages/web/src/helpers/getSplitStyles.tsx [hunk 1]
```diff
@@ -782,6 +782,15 @@ export const getSplitStyles: StyleSplitter = (
                 const defaultVal = animatableDefaults[pkey]
                 mergeStyle(styleState, pkey, defaultVal)
               }
+              if (styleProps.supportsZeroRenderPseudosForAnimations) {
```
Comment (nandorojo) [inline packages/web/src/helpers/getSplitStyles.tsx:785]: @natew would be good to get your review on this section to make sure it's correct. It worked for me in sandbox, but I don't know if I missed something.

File: packages/web/src/createComponent.tsx [hunk 2]
```diff
@@ -767,7 +790,6 @@ export function createComponent<
                   setStateShallow({
                     press: true,
                     pressIn: true,
-                    hover: false,
```
Comment (nandorojo) [inline packages/web/src/createComponent.tsx:770]: Removed this to fix: https://discord.com/channels/909986013848412191/1136717168545906788

File: packages/animations-moti/src/createAnimations.tsx [hunk 3]
```diff
@@ -97,15 +98,20 @@ export function createAnimations<A extends Record<string, MotiTransition>>(
...
+      })
       const animationKey = Array.isArray(props.animation)
         ? props.animation[0]
         : props.animation
 
       let animate: Object | undefined
       let dontAnimate: Object | undefined
 
-      const animateOnly = props.animateOnly || ['transform', 'opacity']
+      const animateOnly = props.animateOnly
```
Comment (nandorojo) [inline packages/animations-moti/src/createAnimations.tsx:114]: I can put `animateOnly` back if we want, but it felt unnecessary to have there. lmk.

File: packages/animations-moti/src/createAnimations.tsx [hunk 4]
```diff
@@ -158,6 +182,18 @@ export function createAnimations<A extends Record<string, MotiTransition>>(
 
       return {
         style: [dontAnimate, moti.style],
+        updatePseudoState: useMemo(
+          // useCallback doesn't infer param types of optional functions 🤷‍♂️
+          // i tried: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64606
+          // - fernando
+          () => (next) => {
+            pseudoState.value = {
```
Comment (kaceycleveland) [inline packages/animations-moti/src/createAnimations.tsx:190]: Dunno if it matters too much but would `Object.assign` help here performance wise? I saw Nate had some issues with spread operators a couple days ago too.

File: packages/animations-moti/src/createAnimations.tsx [hunk 5]
```diff
@@ -132,16 +138,34 @@ export function createAnimations<A extends Record<string, MotiTransition>>(
...
+        const { hover: hovered, press: pressed, focus: focused } = pseudoState.value
+
+        if (hovered && pseudos?.hoverStyle) {
+          final = Object.assign(final, pseudos.hoverStyle)
+        }
+
+        if (pressed && pseudos?.pressStyle) {
+          final = Object.assign(final, pseudos.pressStyle)
+        }
+
```
Comment (nandorojo) [inline packages/animations-moti/src/createAnimations.tsx:153]: not sure if this is the right order or not

File: packages/animations-moti/src/createAnimations.tsx (hunk 4 above)
Comment (nandorojo) [reply packages/animations-moti/src/createAnimations.tsx:190]: could do that if it helps. would need to create a new object instance either way

File: packages/web/src/createComponent.tsx [hunk 6]
```diff
@@ -549,6 +555,23 @@ export function createComponent<
         staticConfig,
       })
 
+      updatePseudoStateForAnimations = animations?.updatePseudoState
+
+      setStateShallow = useCallback(
```
Comment (natew) [inline packages/web/src/createComponent.tsx:560]: no need for callback as it doesnt go into any memo type thing

File: packages/web/src/createComponent.tsx [hunk 7]
```diff
@@ -549,6 +555,23 @@ export function createComponent<
...
+        (next) => {
+          if ('unmounted' in next) {
+            // unmounted is never called on the pseudo side
+            // it would be best if unmounted lived in its own state variable to avoid this level of mixing probs
```
Comment (natew) [inline packages/web/src/createComponent.tsx:564]: everyhting is sensitive to perf so try and avoid as many hooks as possible

File: packages/web/src/createComponent.tsx [hunk 8]
```diff
@@ -549,6 +555,23 @@ export function createComponent<
         staticConfig,
       })
 
+      updatePseudoStateForAnimations = animations?.updatePseudoState
+
+      setStateShallow = useCallback(
+        (next) => {
+          if ('unmounted' in next) {
```
Comment (natew) [inline packages/web/src/createComponent.tsx:562]:
```suggestion
          if ('unmounted' in next || ! updatePseudoStateForAnimations) {
```

**Pull request closed without merging**